File: agent-machine/scripts/knowledge_type.py
#!/usr/bin/env python3
"""
knowledge_type — classify questions by the 7 ARC knowledge types (Boratko et al. 2018,
"A Systematic Classification of Knowledge, Reasoning, and Context within the ARC Dataset").

Each type routes to the SOLVER it depends on — this is the question-type router that sits
beside the domain router (domain = what field; knowledge-type = what kind → which method):

  Definition / BasicFacts / Purpose  → golden RETRIEVAL (lookup)
  Causes & Processes                 → the multi-hop CHAINER
  Algebraic                          → the VERIFIED-COMPUTE engine
  Experiments                        → experimental reasoning (to build)
  Physical Model                     → spatial/kinematic reasoning (to build)

Multi-label (the paper finds ~1.4 types/question). Cheap rule-based v0 off the type
definitions; swap a trained head in later. Run over the MMLU bank to see the type mix.

Run:  python3 scripts/knowledge_type.py
"""
import os, sys, re, json
from collections import Counter

UNIT = re.compile(r'\d\s*(m/s|kg|mol|N\b|J\b|W\b|V\b|A\b|Hz|cm|mm|km|°|K\b|Pa|ohm|Ω|g\b|L\b|eV|nm|watt|volt|joule|gram|meter|second)', re.I)
NUM = re.compile(r'\b\d+(\.\d+)?\b')

# type -> (solver, signal patterns)
TYPES = {
    'Definition':       ('retrieve', [r'\bis called\b', r'\bterm for\b', r'\bdefined as\b', r'\bthe name (for|of)\b',
                                       r'\bbest describes\b', r'\bdefinition of\b', r'\brefers to\b']),
    'BasicFacts':       ('retrieve', [r'\bhow many\b', r'\bwhich (of the )?(following )?(element|compound|gas|metal|organ|planet|part)\b',
                                       r'\bwhat (is|are) the\b', r'\bmade (up )?of\b', r'\bconsists? of\b']),
    'CausesProcesses':  ('chain',    [r'\bfirst step\b', r'\bprocess\b', r'\bsequence\b', r'\bstages?\b', r'\bcycle\b',
                                       r'\bwhat happens (when|after|next|if)\b', r'\bin order\b', r'\bsteps?\b',
                                       r'\bleads? to\b', r'\bresults? in\b', r'\bcauses?\b',
                                       # concept-application (evolution/biology conceptual reasoning)
                                       r'\bexemplif', r'\bexample of\b', r'\billustrat', r'\bdemonstrat',
                                       r'\b(homolog|analog|convergent|divergent|vestigial)\w*\b', r'\bbest (describes|explains)\b']),
    'Purpose':          ('retrieve', [r'\bfunction of\b', r'\bpurpose of\b', r'\brole of\b', r'\bused (to|for)\b',
                                       r'\bwhy (do|does|are|is)\b', r'\bin order to\b', r'\bhelps? (to )?\b']),
    'Algebraic':        ('compute',  [r'\bcalculate\b', r'\bhow much\b', r'\bhow far\b', r'\bhow fast\b', r'\bhow long\b',
                                       r'\bwhat is the (value|magnitude|force|velocity|energy|current|mass|speed|frequency|resistance|acceleration|momentum|pressure|power|charge|wavelength|volume|density|work)\b',
                                       r'=', r'\bsolve\b', r'\bequation\b']),
    'Experiments':      ('experiment', [r'\bexperiment', r'\bhypothes', r'\bcontrol(led| group| variable)?\b', r'\bindependent variable\b',
                                        r'\bscientists?\b', r'\bmeasure(ment)?\b', r'\bobserv', r'\btest(ed|ing)?\b', r'\bdata\b', r'\btrial\b']),
    'PhysicalModel':    ('spatial', [r'\bmoves?\b', r'\bmoving\b', r'\bcollid', r'\bdirection\b', r'\bdistance\b', r'\bpath\b',
                                     r'\bwhat (most likely )?happens when\b', r'\bair mass\b', r'\borbit', r'\bgravit', r'\bfriction\b']),
}


def classify(q):
    text = q.lower()
    hits = []
    for t, (_solver, sigs) in TYPES.items():
        if any(re.search(s, text) for s in sigs):
            hits.append(t)
    if 'Algebraic' not in hits and (UNIT.search(q) or (re.search(r'\bwhat is the\b', text) and NUM.search(q))):
        hits.append('Algebraic')
    return hits or ['BasicFacts']   # default to a factual lookup


def _batch():
    """Read JSONL {id, question, choices}; emit {id, types, solver} so the exam can classify each
    question BEFORE approaching it and route to the right method (the 'understand first' step)."""
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            q = json.loads(line)
            types = classify(q['question'])
            # pick the highest-PRIORITY solver among matched types, not types[0] — a question that
            # is both 'which…' (BasicFacts) and computational must route to compute, not retrieve.
            solvers = {TYPES[t][0] for t in types}
            solver = next((s for s in ('compute', 'chain', 'spatial', 'experiment', 'retrieve') if s in solvers), 'retrieve')
        except Exception:
            q, types, solver = {}, ['BasicFacts'], 'retrieve'
        print(json.dumps({'id': q.get('id'), 'types': types, 'solver': solver}), flush=True)


def main():
    if '--batch' in sys.argv:
        return _batch()
    bank = json.load(open(os.path.expanduser('~/.noetica/corpus/benchmarks/mmlu_stem.json')))
    type_count = Counter(); solver_count = Counter(); n = 0; multi = 0
    by_subject = {}
    for subj, items in bank.items():
        sc = Counter()
        for it in items:
            n += 1
            types = classify(it['question'])
            if len(types) > 1:
                multi += 1
            for t in types:
                type_count[t] += 1; sc[t] += 1
                solver_count[TYPES[t][0]] += 1
        by_subject[subj] = sc

    print(f"# knowledge-type classification — {n} MMLU questions · {multi/n:.0%} multi-label "
          f"(paper: ~1.4 types/q)\n")
    print("  KNOWLEDGE TYPE      questions   share   → solver")
    print("  " + "-"*54)
    for t, c in type_count.most_common():
        print(f"  {t:18} {c:>7}   {100*c//n:>4}%   → {TYPES[t][0]}")
    print(f"\n  SOLVER ROUTING (where the type-router sends each question):")
    for s, c in solver_count.most_common():
        print(f"    {s:12} {c:>6}  ({100*c//sum(solver_count.values()):>2}%)")
    print(f"\n  by subject (dominant knowledge type):")
    for subj, sc in by_subject.items():
        top = sc.most_common(1)[0] if sc else ('-', 0)
        print(f"    {subj:26} {top[0]} ({100*top[1]//len(bank[subj])}%)")


if __name__ == '__main__':
    main()

File: agent-machine/lib/knowledge-type.ts
/**
 * knowledge-type — classify a question by the 7 ARC knowledge types (Boratko et al. 2018) and route
 * it to the SOLVER it depends on, BEFORE answering. This is the "understand first" step that decides
 * whether a turn is LOOKUP-dominated (retrieve), COMPUTE-dominated (verified sympy), or MODEL-dominated
 * (reason/generate) — the per-question fine gate beside the action-polarity coarse gate.
 *
 * Ported from scripts/knowledge_type.py (was bench-only) so the DIALOGUE can classify every turn.
 * Validated by the 2025-26 medical-QA literature: ~33% of medical QA is reasoning-bound and ~67% is
 * knowledge-bound; RAG helps the lookup bucket, reasoning the model bucket, and uncertainty-gating
 * decides between them (Med-Gemini, MedRAG/MIRAGE, "Disentangling Reasoning & Knowledge" 2505.11462).
 */

export type KnowledgeType = 'Definition' | 'BasicFacts' | 'Purpose' | 'CausesProcesses' | 'Algebraic' | 'Experiments' | 'PhysicalModel'
export type Solver = 'retrieve' | 'compute' | 'chain' | 'spatial' | 'experiment'
export type Dominance = 'lookup' | 'compute' | 'model'
export interface KnowledgeClass { types: KnowledgeType[]; solver: Solver; dominance: Dominance }

const TYPES: Record<KnowledgeType, { solver: Solver; sigs: RegExp[] }> = {
  Definition: { solver: 'retrieve', sigs: [/\bis called\b/i, /\bterm for\b/i, /\bdefined as\b/i, /\bthe name (for|of)\b/i, /\bbest describes\b/i, /\bdefinition of\b/i, /\brefers to\b/i, /\bwhat (is|are) (a|an|the)\b/i, /\bwhat'?s (a|an|the)\b/i] },
  BasicFacts: { solver: 'retrieve', sigs: [/\bhow many\b/i, /\bwhich (of the )?(following )?(element|compound|gas|metal|organ|planet|part|drug|enzyme|hormone)\b/i, /\bwhat (is|are) the\b/i, /\bmade (up )?of\b/i, /\bconsists? of\b/i] },
  Purpose: { solver: 'retrieve', sigs: [/\bfunction of\b/i, /\bpurpose of\b/i, /\brole of\b/i, /\bused (to|for)\b/i, /\bwhy (do|does|are|is)\b/i, /\bin order to\b/i, /\bhelps? (to )?\b/i] },
  // NOTE: "best describes" lives in Definition (lookup), NOT here — a "which best describes X" question
  // is a definition lookup, and since 'chain' outranks 'retrieve' (line ~52) keeping it here would
  // mis-route every such lookup to model-dominance. "best explains" IS causal, so it stays.
  CausesProcesses: { solver: 'chain', sigs: [/\bfirst step\b/i, /\bprocess\b/i, /\bsequence\b/i, /\bstages?\b/i, /\bcycle\b/i, /\bwhat happens (when|after|next|if)\b/i, /\bin order\b/i, /\bsteps?\b/i, /\bleads? to\b/i, /\bresults? in\b/i, /\bcauses?\b/i, /\bexemplif/i, /\bexample of\b/i, /\billustrat/i, /\bdemonstrat/i, /\b(homolog|analog|convergent|divergent|vestigial)\w*\b/i, /\bbest explains\b/i, /\bmost likely (cause|diagnosis|explanation)\b/i] },
  // `=` must be a REAL equality (digit/var on the left, signed number on the right) — a bare /=/ fired on
  // ==, <=, !=, and any stray '=' in prose, force-routing non-numeric questions to the compute solver.
  Algebraic: { solver: 'compute', sigs: [/\bcalculate\b/i, /\bhow much\b/i, /\bhow far\b/i, /\bhow fast\b/i, /\bhow long\b/i, /\bwhat is the (value|magnitude|force|velocity|energy|current|mass|speed|frequency|resistance|acceleration|momentum|pressure|power|charge|wavelength|volume|density|work|probability)\b/i, /[\dxyz)]\s*=\s*[-+(]?\s*\d/i, /\bsolve\b/i, /\bequation\b/i] },
  Experiments: { solver: 'experiment', sigs: [/\bexperiment/i, /\bhypothes/i, /\bcontrol(led| group| variable)?\b/i, /\bindependent variable\b/i, /\bscientists?\b/i, /\bmeasure(ment)?\b/i, /\bobserv/i, /\btest(ed|ing)?\b/i, /\bdata\b/i, /\btrial\b/i] },
  PhysicalModel: { solver: 'spatial', sigs: [/\bmoves?\b/i, /\bmoving\b/i, /\bcollid/i, /\bdirection\b/i, /\bdistance\b/i, /\bpath\b/i, /\bwhat (most likely )?happens when\b/i, /\bair mass\b/i, /\borbit/i, /\bgravit/i, /\bfriction\b/i] },
}

const UNIT = /\d\s*(m\/s|kg|mol|N\b|J\b|W\b|V\b|A\b|Hz|cm|mm|km|°|K\b|Pa|ohm|Ω|g\b|L\b|eV|nm|mg|ml|mmHg|watt|volt|joule|gram|meter|second)/i
const NUM = /\b\d+(\.\d+)?\b/

const SOLVER_DOMINANCE: Record<Solver, Dominance> = {
  retrieve: 'lookup', compute: 'compute', chain: 'model', spatial: 'model', experiment: 'model',
}

/** Classify a question by knowledge type → solver → lookup/compute/model dominance. Multi-label. */
export function classifyKnowledge(question: string): KnowledgeClass {
  const hits: KnowledgeType[] = []
  for (const t of Object.keys(TYPES) as KnowledgeType[]) {
    if (TYPES[t].sigs.some((s) => s.test(question))) hits.push(t)
  }
  // numeric fallback → Algebraic ONLY on a real physical quantity (unit). A bare number — an age
  // ("55-year-old"), a year, patient demographics in a clinical vignette — must NOT route to compute;
  // that mislabels a reasoning question. (Explicit math — "=", solve, calculate, probability — is
  // already a typed Algebraic signal above; hyphens in compounds are NOT minus operators.)
  if (!hits.includes('Algebraic') && UNIT.test(question)) {
    hits.push('Algebraic')
  }
  const types = hits.length ? hits : (['BasicFacts'] as KnowledgeType[])
  // pick the highest-PRIORITY solver among matched types — a question that is BOTH "which…" and
  // computational must route to compute, not retrieve.
  const solvers = new Set(types.map((t) => TYPES[t].solver))
  const solver: Solver = (['compute', 'chain', 'spatial', 'experiment', 'retrieve'] as Solver[]).find((s) => solvers.has(s)) ?? 'retrieve'
  return { types, solver, dominance: SOLVER_DOMINANCE[solver] }
}
